scripts/igmp_query_injector.py

```diff
@@ -0,0 +1,459 @@
+
+    def state_is_expected(self):
+        return XenstoreUtil.read(self.vif_state_path) == self.expect_state
+
+
+class IGMPQueryGenerator(object):
+    """IGMP Query generator.
+    Generate IGMP Query packet with/without vlan
+    """
+    def __init__(self, dst_mac, vlanid, max_resp_time):
+        """
+        :param dst_mac: Destination mac address of this IGMP query packet
+        :param vlanid: Vlan ID of this packet. `-1` means no vlan
+        :param max_resp_time: Max response time of IGMP query. Unit is 100ms
+        """
+        self.src_mac = '00:00:00:00:00:00'
+        self.dst_mac = dst_mac
+        self.vlanid = vlanid
+        self.max_resp_time = max_resp_time
+
+    def parse_mac_address(self, mac):
+        ret = []
+        for x in mac.split(':'):
+            ret.append(int(x, 16))
+
+        return ret
+
+    def create_ether_layer(self):
+        ret = []
+        for mac in (self.dst_mac, self.src_mac):
+            ret.extend(self.parse_mac_address(mac))
+
+        if self.vlanid == -1:
+            type_field = [0x08, 0x00]
+        else:
+            type_field = [0x81, 0x00]
+
+        ret.extend(type_field)
+        return ret
+
+    def create_vlan_layer(self):
+        if self.vlanid == -1:
+            return []
+        else:
+            return [0x20 | (self.vlanid >> 8), self.vlanid & 0xff, 0x08, 0x00]
+
+    def create_ip_layer(self):
+        """
+        IP Type: IPv4 (0x0800)
+        Version: 4
+        Total length: 32
+        TTL: 1
+        Source IP: 0.0.0.0
+        Destination IP: 224.0.0.1 (e0:00:00:01)
+        """
+        return [
+            0x46, 0x00, 0x00, 0x20, 0x00, 0x01,
+            0x00, 0x00, 0x01, 0x02, 0x44, 0xd6,
+            0x00, 0x00, 0x00, 0x00, 0xe0, 0x00,
+            0x00, 0x01, 0x94, 0x04, 0x00, 0x00,
+        ]
+
+    def create_igmp_layer(self):
+        """
+        IGMP Version: 2
+        IGMP Type: IGMP Query (0x11)
+        Max Resp Time: self.max_resp_time
+        Multicast Address: 0.0.0.0
+        """
+        msg_type = 0x11
+        # MAX_RESP_TIME filed unit is 100ms
+        max_resp_time = self.max_resp_time / 100
+        group_address = [0x00, 0x00, 0x00, 0x00]
+        chksum = self._calc_igmp_chksum(msg_type, max_resp_time, group_address)
+        ret = [msg_type, max_resp_time] + chksum + group_address
+        return ret
+
+    def _calc_igmp_chksum(self, msg_type, max_resp_time, group_address):
+        msg = [msg_type, max_resp_time, 0x00, 0x00] + group_address
+        chksum = socket.htons(self._checksum(msg))
+        return [(chksum & 0xff00) >> 8, chksum & 0xff]
+
+    def _checksum(self, msg):
+        def carry_around_add(a, b):
+            c = a + b
+            return (c & 0xffff) + (c >> 16)
+
+        s = 0
+        for i in range(0, len(msg), 2):
+            w = msg[i] + (msg[i+1] << 8)
+            s = carry_around_add(s, w)
+        return ~s & 0xffff
+
+    def generate(self):
+        ether_layer = self.create_ether_layer()
```

Comment:
We should consider whether to import scapy into dom0 and use it for this.

Reply:
I have considered this but after checked the contents of `site-packages` I found Dom0 just installed limited 3rd-party libraries. For this feature, we just need construct the IGMP query message with only 1 field(Max Resp Time) customized, so I think it's unnecessary to introduce `scapy`  into Dom0.